### PR TITLE
Positioning

### PR DESCRIPTION
**Pull request by bit-creator**

### CODE REVIEW

File: object3d.cpp [hunk 1]
```diff
@@ -24,3 +24,67 @@ Object3D& Object3D::operator =(const Object3D& oth) noexcept
...
+colour_t Object3D::getColor() const noexcept
+{ return _colour; }
+
+void Object3D::setScale(const glm::vec3& scale) noexcept
+{ _scale = scale; }
+
+glm::vec3 Object3D::getScale() const noexcept
+{ return _scale; }
+
+void Object3D::setRotate(const Quaternion& rotate) noexcept
```
Comment (LuxoftMentor) [inline object3d.cpp]: Change to setRotate(vec3 axis, float angle)

File: quad.cpp [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+    glm::mat3 mat;
+
+    glm::mat3 mRotate = generateRotateMatrix(getRotate());
+
+    glm::mat3x3 mScale(scale.x, 0.,      0.,
+                   0.,      scale.y, 0.,
+                   0.,      0.,      scale.z);
+
+    mat = mRotate * mScale;
```
Comment (LuxoftMentor) [inline quad.cpp]: Move to mat3 Object3D::getModelMat();

File: shaders/primitive/primitive.vert.glsl [hunk 3]
```diff
@@ -1,10 +1,11 @@
 #version 460 core
 
-layout(location = 0)
+layout(location = 0) in vec3 pos;
```
Comment (LuxoftMentor) [inline shaders/primitive/primitive.vert.glsl]: Rename to aCoord;

File: shaders/primitive/primitive.vert.glsl [hunk 4]
```diff
@@ -1,10 +1,11 @@
...
 
-in vec3 pos;
+uniform mat3 mat;
+uniform vec3 offset;
```
Comment (LuxoftMentor) [inline shaders/primitive/primitive.vert.glsl]: Rename to uPosition

File: shaders/primitive/primitive.vert.glsl [hunk 5]
```diff
@@ -1,10 +1,11 @@
 #version 460 core
 
-layout(location = 0)
+layout(location = 0) in vec3 pos;
 
-in vec3 pos;
+uniform mat3 mat;
```
Comment (LuxoftMentor) [inline shaders/primitive/primitive.vert.glsl]: Rename to uModelMat;

File: object3d.h [hunk 6]
```diff
@@ -18,9 +18,10 @@
 
 #include "GL/shaderprogram.h"
 
-using colour_t = glm::vec3;
-using normal_t = glm::vec3;
-using vertex_t = std::shared_ptr<glm::vec3>;
+using colour_t   = glm::vec4;
+using normal_t   = glm::vec3;
+using vertex_t   = std::shared_ptr<glm::vec3>;
```
Comment (LuxoftMentor) [inline object3d.h]: Don`t use smart pointers with simple types.

File: engine.cpp [hunk 7]
```diff
@@ -1,9 +1,14 @@
 #include "engine.h"
 
-const Engine& Engine::engine() noexcept
+Engine& Engine::engine() noexcept
 { static Engine engine; return engine; }
 
-void Engine::run(const Object3D* tr, const Window& window) const noexcept
+void Engine::addEventListener(EventPointer event)
```
Comment (LuxoftMentor) [inline engine.cpp]: Rename event to eventListener

File: main.cpp [hunk 8]
```diff
@@ -2,27 +2,78 @@
...
 {
     system("clear");
 
+    // ShaderTree::get().initShaderTree("shaders");
+
+    auto& eng = Engine::engine();
+
     vertex_t vert_1 = std::make_shared<glm::vec3>(glm::vec3(-0.5f, -0.5f, 0.0f));
     vertex_t vert_2 = std::make_shared<glm::vec3>(glm::vec3(0.5f, -0.5f, 0.0f));
-    vertex_t vert_3 = std::make_shared<glm::vec3>(glm::vec3(0.0f, 0.5f, 0.0f));
```
Comment (LuxoftMentor) [inline main.cpp:19]: Remove shared pointers. Use just vec3

File: quad.h [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+ * @date 2020-11-21
+ * 
+ * @copyright Copyright (c) 2020
+ * 
+ */
+
+#ifndef QUAD_H
+#define QUAD_H
+
+#include "triangle.h"
```
Comment (LuxoftMentor) [inline quad.h]: Avoid dependency

File: triangle.cpp [hunk 10]
```diff
@@ -70,7 +70,21 @@ void Triangle::render(const ShaderProgram& shader) const noexcept
...
+    auto scale = getScale();
+
+    glm::mat3x3 ms(scale.x, 0.,      0.,
+                   0.,      scale.y, 0.,
+                   0.,      0.,      scale.z);
+
+
+    // glm::mat3x3 mr(  )
+    
+
```
Comment (LuxoftMentor) [inline triangle.cpp]: Copy implementation from quad

**Review by LuxoftMentor — CHANGES_REQUESTED** [review]

File: object3d.cpp (hunk 1 above)
Comment (bit-creator) [reply object3d.cpp]: problem fixed

**Pull request merged**